### PR TITLE
[SPARK-18697][BUILD] Upgrade sbt plugins

### PR DESCRIPTION
**Pull request by weiqingy**
## What changes were proposed in this pull request?

This PR is to upgrade sbt plugins. The following sbt plugins will be upgraded:
```
sbt-assembly: 0.11.2 -> 0.14.3
sbteclipse-plugin: 4.0.0 -> 5.0.1
sbt-mima-plugin: 0.1.11 -> 0.1.12
org.ow2.asm/asm: 5.0.3 -> 5.1 
org.ow2.asm/asm-commons: 5.0.3 -> 5.1 
```
All other plugins are up-to-date.

## How was this patch tested?
Pass the Jenkins build.

### CODE REVIEW

**Comment by weiqingy** [comment]
The references to upgrade sbt-assembly as following:  
https://github.com/sbt/sbt-assembly/blob/master/Migration.md
https://github.com/sbt/sbt-assembly/blob/0.12.0/src/main/scala/sbtassembly/AssemblyKeys.scala

**Comment by weiqingy** [comment]
It failed the unit test: `org.apache.spark.sql.streaming.StreamingQueryListenerSuite.single listener, check trigger events are generated correctly`. However, I just ran it locally, and it passed. so will re-trigger the test.

**Comment by weiqingy** [comment]
Retest this please.

File: project/SparkBuild.scala [hunk 1]
```diff
@@ -596,19 +596,17 @@ object Hive {
...
 object Assembly {
-  import sbtassembly.AssemblyUtils._
-  import sbtassembly.Plugin._
-  import AssemblyKeys._
+  import sbtassembly.AssemblyPlugin.autoImport._
 
   val hadoopVersion = taskKey[String]("The version of hadoop that spark is compiled against.")
 
-  lazy val settings = assemblySettings ++ Seq(
+  lazy val settings = Seq(
```
Comment (srowen) [inline project/SparkBuild.scala]: I don't know enough about the SBT changes to evaluate these, but you seem like you know what you're doing. :) Just checking that this change was on purpose, to remove assemblySettings.

**Review by srowen — APPROVED** [review]
Seems OK for master.

**Comment by srowen** [comment]
Merged to master

**Comment by weiqingy** [comment]
Thanks for the review, @srowen .

**Comment by zsxwing** [comment]
Looks like this one breaks Scala 2.10 compile jobs: https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Compile/job/spark-master-compile-sbt-scala-2.10/3301/

**Comment by srowen** [comment]
Shoot OK, I'll revert it. Sorry @weiqingy . Maybe you are expert enough to know how to fix that?

**Comment by zsxwing** [comment]
I don't know how to fix it...

**Comment by srowen** [comment]
Maybe @weiqingy does I mean ... at the least maybe some of these plugins can be safely updated. Or soon I think we'll flip off Scala 2.10 support in master and then this can be merged.

**Comment by weiqingy** [comment]
Thanks, @srowen @zsxwing . My bad. I am looking into it.

**Pull request closed without merging**